rust/geoarrow/src/io/csv/reader.rs

```diff
@@ -1,57 +1,189 @@
         }
+        if let Some(terminator) = self.terminator {
+            format = format.with_terminator(terminator as u8);
+        }
+        if let Some(comment) = self.comment {
+            format = format.with_comment(comment as u8);
+        }
+
+        format
     }
 }
 
 impl Default for CSVReaderOptions {
     fn default() -> Self {
-        Self::new(Default::default(), 65_536)
+        Self {
+            coord_type: Default::default(),
+            batch_size: 65_536,
+            geometry_column_name: Default::default(),
+            has_header: Default::default(),
+            max_records: Default::default(),
+            delimiter: Default::default(),
+            escape: Default::default(),
+            quote: Default::default(),
+            terminator: Default::default(),
+            comment: Default::default(),
+        }
     }
 }
 
+/// Infer a CSV file's schema
+/// Returns (Schema, records_read, geometry column name)
+///
+/// Note that the geometry column in the Schema is still left as a String.
+pub fn infer_csv_schema(
+    reader: impl Read,
+    options: &CSVReaderOptions,
+) -> Result<(Schema, usize, String)> {
+    let format = options.to_format();
+    let (schema, records_read) = format.infer_schema(reader, options.max_records)?;
+
+    let geometry_col_name = find_geometry_column(
+        &schema,
+        options.geometry_column_name.as_ref().map(|x| x.as_str()),
+    )?;
+
+    Ok((schema, records_read, geometry_col_name))
+}
+
 /// Read a CSV file to a Table
+///
+/// Note that this is Read and not Read + Seek. This means that you must infer the schema yourself
+/// before calling this function. This allows using with objects that are only `Read` in the case
+/// when you already know the file's schema.
 pub fn read_csv<R: Read>(
-    reader: R,
-    geometry_column_name: &str,
+    mut reader: R,
+    schema: SchemaRef,
     options: CSVReaderOptions,
 ) -> Result<Table> {
-    let mut csv = CsvReader::new(geometry_column_name, reader);
-    let table_builder_options = GeoTableBuilderOptions::new(
-        options.coord_type,
-        true,
-        Some(options.batch_size),
-        None,
-        None,
-        Default::default(),
-    );
-    let mut geo_table = GeoTableBuilder::<MixedGeometryStreamBuilder>::new_with_options(
-        Dimension::XY,
-        table_builder_options,
-    );
-    csv.process(&mut geo_table)?;
-    geo_table.finish()
+    let geometry_column_name = find_geometry_column(
+        schema.as_ref(),
+        options.geometry_column_name.as_ref().map(|x| x.as_str()),
+    )?;
+    let mut builder = ReaderBuilder::new(schema)
+        .with_format(options.to_format())
+        .with_batch_size(options.batch_size);
+
+    // Transform to output schema
+    // TODO:
+    let output_schema = Schema::new(vec![]);
+
+    // TODO:
+    let geometry_column_index = 0;
+
+    let reader = builder.build(reader)?;
+    for record_batch in reader {
```

Comment:
todo: change this loop to an iterator so that we can return a `RecordBatchIterator` (which implements `RecordBatchReader`) to an `arro3.core.RecordBatchReader`